File: 3462.maximum-sum-with-at-most-k-elements.py
#
# @lc app=leetcode id=3462 lang=python3
#
# [3462] Maximum Sum With at Most K Elements
#

# @lc code=start
from typing import List
class Solution:
    def maxSum(self, grid: List[List[int]], limits: List[int], k: int) -> int:
        n = len(grid)
        dp = [0] + [-float('inf')] * k  # dp[j]: max sum for picking j elements
        for row, limit in zip(grid, limits):
            m = len(row)
            pick_limit = min(limit, m)
            sorted_row = sorted(row, reverse=True)
            prefix = [0]
            for i in range(pick_limit):
                prefix.append(prefix[-1] + sorted_row[i])
            ndp = dp[:]
            for t in range(1, len(prefix)):
                for j in range(k, t - 1, -1):
                    if dp[j - t] != -float('inf'):
                        ndp[j] = max(ndp[j], dp[j - t] + prefix[t])
            dp = ndp
        return max(dp)
# @lc code=end

File: 3462.maximum-sum-with-at-most-k-elements.java
#
# @lc app=leetcode id=3462 lang=java
#
# [3462] Maximum Sum With at Most K Elements
#
# @lc code=start
class Solution {
    public long maxSum(int[][] grid, int[] limits, int k) {
        int n = grid.length, m = grid[0].length;
        long[] dp = new long[k + 1];
        for (int i = 0; i < n; i++) {
            int[] row = grid[i];
            Arrays.sort(row);
            int lim = limits[i];
            int take = Math.min(lim, row.length);
            long[] prefix = new long[take + 1];
            for (int j = 1; j <= take; j++) {
                prefix[j] = prefix[j - 1] + row[row.length - j];
            }
            long[] ndp = dp.clone();
            for (int cnt = 0; cnt <= take; cnt++) {
                for (int j = k; j >= cnt; j--) {
                    ndp[j] = Math.max(ndp[j], dp[j - cnt] + prefix[cnt]);
                }
            }
            dp = ndp;
        }
        long ans = 0;
        for (int i = 0; i <= k; i++) ans = Math.max(ans, dp[i]);
        return ans;
    }
}
# @lc code=end